

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>BrainNetAnno.transcriptome_pls_cge &mdash; BrainNetAnno  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            BrainNetAnno
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <!-- Local TOC -->
              <div class="local-toc"></div>
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">BrainNetAnno</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">BrainNetAnno.transcriptome_pls_cge</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for BrainNetAnno.transcriptome_pls_cge</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">PLS analysis for transcriptome gene contributions against FC deviations.</span>

<span class="sd">Provides reusable functions to select PLS components via CV, run PLS, compute</span>
<span class="sd">permutation p-values for weights, plotting utilities, and high-level pipeline.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Optional</span>
<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">zscore</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="kn">from</span> <span class="nn">.utils</span> <span class="kn">import</span> <span class="n">load_table_generic</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(asctime)s</span><span class="s2"> </span><span class="si">%(levelname)s</span><span class="s2"> </span><span class="si">%(name)s</span><span class="s2">: </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>

<span class="k">try</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">BrainNetAnno.utils</span> <span class="kn">import</span> <span class="p">(</span>
        <span class="n">select_optimal_components</span><span class="p">,</span>
        <span class="n">permutation_explained_variance</span><span class="p">,</span>
    <span class="p">)</span>
<span class="k">except</span> <span class="ne">ModuleNotFoundError</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">sys</span><span class="o">,</span> <span class="nn">os</span>
    <span class="n">src_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="vm">__file__</span><span class="p">),</span> <span class="s1">&#39;..&#39;</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">src_dir</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">src_dir</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="kn">from</span> <span class="nn">utils</span> <span class="kn">import</span> <span class="p">(</span>
            <span class="n">select_optimal_components</span><span class="p">,</span>
            <span class="n">permutation_explained_variance</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">except</span> <span class="ne">ModuleNotFoundError</span><span class="p">:</span>
        <span class="n">project_root</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">src_dir</span><span class="p">,</span> <span class="s1">&#39;..&#39;</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">project_root</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">project_root</span><span class="p">)</span>
        <span class="kn">from</span> <span class="nn">BrainNetAnno.utils</span> <span class="kn">import</span> <span class="p">(</span>
            <span class="n">select_optimal_components</span><span class="p">,</span>
            <span class="n">permutation_explained_variance</span><span class="p">,</span>
        <span class="p">)</span>

<span class="c1"># -----------------------------</span>
<span class="c1"># Data reading</span>
<span class="c1"># -----------------------------</span>

<div class="viewcode-block" id="read_data_from_files">
<a class="viewcode-back" href="../../BrainNetAnno.html#BrainNetAnno.transcriptome_pls_cge.read_data_from_files">[docs]</a>
<span class="k">def</span> <span class="nf">read_data_from_files</span><span class="p">(</span><span class="n">t_values_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">gene_expression_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Read FC deviation (t_values) and gene contribution matrices, align by Region_Pair.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    t_values_file : str</span>
<span class="sd">        Path to CSV containing a square matrix of FC deviations (no header).</span>
<span class="sd">    gene_expression_file : str</span>
<span class="sd">        Path to CSV containing gene contributions with a &#39;Region_Pair&#39; column.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    Tuple[pd.DataFrame, np.ndarray, np.ndarray, np.ndarray]</span>
<span class="sd">        (df_pairs_values, t_values_vector, gene_matrix_aligned, gene_names_array)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">t_values</span> <span class="o">=</span> <span class="n">load_table_generic</span><span class="p">(</span><span class="n">t_values_file</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">gene_df</span> <span class="o">=</span> <span class="n">load_table_generic</span><span class="p">(</span><span class="n">gene_expression_file</span><span class="p">)</span>

    <span class="c1"># Upper triangle non-zero entries from t_values</span>
    <span class="n">coords</span> <span class="o">=</span> <span class="p">[(</span><span class="n">c</span><span class="p">,</span> <span class="n">r</span><span class="p">)</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">t_values</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">r</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">t_values</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="k">if</span> <span class="n">t_values</span><span class="o">.</span><span class="n">iat</span><span class="p">[</span><span class="n">r</span><span class="p">,</span> <span class="n">c</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">]</span>
    <span class="n">formatted</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">r</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">c</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">coords</span><span class="p">]</span>

    <span class="c1"># Align gene_df by Region_Pair</span>
    <span class="n">filtered</span> <span class="o">=</span> <span class="n">gene_df</span><span class="p">[</span><span class="n">gene_df</span><span class="p">[</span><span class="s2">&quot;Region_Pair&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">formatted</span><span class="p">)]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">gene_names</span> <span class="o">=</span> <span class="n">filtered</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span>
    <span class="n">gene_names</span> <span class="o">=</span> <span class="n">gene_names</span><span class="p">[</span><span class="n">gene_names</span> <span class="o">!=</span> <span class="s2">&quot;Region_Pair&quot;</span><span class="p">]</span>
    <span class="n">gene_matrix</span> <span class="o">=</span> <span class="n">filtered</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Region_Pair&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">values</span>

    <span class="n">df_pairs_values</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">r</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">c</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">t_values</span><span class="o">.</span><span class="n">at</span><span class="p">[</span><span class="n">r</span><span class="p">,</span> <span class="n">c</span><span class="p">])</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">coords</span><span class="p">],</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Region_Pair&quot;</span><span class="p">,</span> <span class="s2">&quot;Value&quot;</span><span class="p">])</span>
    <span class="n">t_vector</span> <span class="o">=</span> <span class="n">df_pairs_values</span><span class="p">[[</span><span class="s2">&quot;Value&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">values</span>

    <span class="k">return</span> <span class="n">df_pairs_values</span><span class="p">,</span> <span class="n">t_vector</span><span class="p">,</span> <span class="n">gene_matrix</span><span class="p">,</span> <span class="n">gene_names</span></div>


<span class="c1"># -----------------------------</span>
<span class="c1"># Select component and genes</span>
<span class="c1"># -----------------------------</span>

<div class="viewcode-block" id="select_best_component_and_genes">
<a class="viewcode-back" href="../../BrainNetAnno.html#BrainNetAnno.transcriptome_pls_cge.select_best_component_and_genes">[docs]</a>
<span class="k">def</span> <span class="nf">select_best_component_and_genes</span><span class="p">(</span><span class="n">explained_variance_ratio</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
                                    <span class="n">p_values</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
                                    <span class="n">weights</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
                                    <span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Select best component (prefer significant) and significant genes by simple z-score permutation.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    explained_variance_ratio : np.ndarray</span>
<span class="sd">        Ratio of variance explained per component.</span>
<span class="sd">    p_values : np.ndarray</span>
<span class="sd">        P-values per component from permutation.</span>
<span class="sd">    weights : np.ndarray</span>
<span class="sd">        X weights matrix from PLS, shape (n_features, n_components).</span>
<span class="sd">    threshold : float, optional</span>
<span class="sd">        Significance threshold, by default 0.05.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    Tuple[int, np.ndarray, np.ndarray, np.ndarray]</span>
<span class="sd">        best_component_index, best_genes_indices, z_scores_per_gene, p_values_per_gene</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">significant_components</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">p_values</span> <span class="o">&lt;</span> <span class="n">threshold</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">significant_components</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">best_component</span> <span class="o">=</span> <span class="n">significant_components</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">explained_variance_ratio</span><span class="p">[</span><span class="n">significant_components</span><span class="p">])]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">best_component</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">explained_variance_ratio</span><span class="p">))</span>

    <span class="n">best_weights</span> <span class="o">=</span> <span class="n">weights</span><span class="p">[:,</span> <span class="n">best_component</span><span class="p">]</span>
    <span class="n">z_scores</span> <span class="o">=</span> <span class="n">zscore</span><span class="p">(</span><span class="n">best_weights</span><span class="p">)</span>
    <span class="n">p_values_genes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">permutation</span><span class="p">(</span><span class="n">z_scores</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">z</span><span class="p">))</span> <span class="k">for</span> <span class="n">z</span> <span class="ow">in</span> <span class="n">z_scores</span><span class="p">])</span>
    <span class="n">best_genes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">p_values_genes</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">best_component</span><span class="p">,</span> <span class="n">best_genes</span><span class="p">,</span> <span class="n">z_scores</span><span class="p">,</span> <span class="n">p_values_genes</span></div>


<span class="c1"># -----------------------------</span>
<span class="c1"># Save and plot</span>
<span class="c1"># -----------------------------</span>

<div class="viewcode-block" id="save_best_genes_to_csv">
<a class="viewcode-back" href="../../BrainNetAnno.html#BrainNetAnno.transcriptome_pls_cge.save_best_genes_to_csv">[docs]</a>
<span class="k">def</span> <span class="nf">save_best_genes_to_csv</span><span class="p">(</span><span class="n">gene_names</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
                           <span class="n">best_genes</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
                           <span class="n">z_scores</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
                           <span class="n">p_values_genes</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span>
                           <span class="n">output_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Save significant genes info to CSV file.&quot;&quot;&quot;</span>
    <span class="n">results</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;Gene Index&#39;</span><span class="p">:</span> <span class="n">gene_names</span><span class="p">[</span><span class="n">best_genes</span><span class="p">],</span>
        <span class="s1">&#39;Original Z-Score&#39;</span><span class="p">:</span> <span class="n">z_scores</span><span class="p">[</span><span class="n">best_genes</span><span class="p">],</span>
        <span class="s1">&#39;P-Value&#39;</span><span class="p">:</span> <span class="n">p_values_genes</span><span class="p">[</span><span class="n">best_genes</span><span class="p">]</span>
    <span class="p">}</span>
    <span class="n">results_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
    <span class="n">results_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;P-Value&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">results_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></div>


<span class="c1"># -----------------------------</span>
<span class="c1"># High-level pipeline</span>
<span class="c1"># -----------------------------</span>

<div class="viewcode-block" id="run_transcriptome_pls_pipeline">
<a class="viewcode-back" href="../../BrainNetAnno.html#BrainNetAnno.transcriptome_pls_cge.run_transcriptome_pls_pipeline">[docs]</a>
<span class="k">def</span> <span class="nf">run_transcriptome_pls_pipeline</span><span class="p">(</span>
    <span class="n">t_values_file_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">gene_expression_file_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">output_best_genes_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">mse_plot_path</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">explained_variance_plot_path</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">max_components</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">15</span><span class="p">,</span>
    <span class="n">cv_splits</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
    <span class="n">random_state</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">42</span><span class="p">,</span>
    <span class="n">n_permutations</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
    <span class="n">n_components</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Run PLS-based transcriptome analysis linking FC deviations to gene contributions.</span>

<span class="sd">    This pipeline reads FC deviation (t-values) and per-connection gene</span>
<span class="sd">    contribution matrices, aligns them by ``Region_Pair``, standardizes inputs,</span>
<span class="sd">    selects the optimal number of PLS components via cross-validated RMSE,</span>
<span class="sd">    estimates permutation-based significance for explained variance and feature</span>
<span class="sd">    weights, saves the significant genes to CSV, and plots RMSE and explained</span>
<span class="sd">    variance curves.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    t_values_file : str</span>
<span class="sd">        Path to the CSV containing a square matrix of FC deviations (no header).</span>
<span class="sd">    gene_expression_file : str</span>
<span class="sd">        Path to the CSV containing per-connection gene contributions with a</span>
<span class="sd">        ``&#39;Region_Pair&#39;`` column.</span>
<span class="sd">    output_best_genes_csv : str</span>
<span class="sd">        Path to save the significant genes table. Parent directories are</span>
<span class="sd">        created automatically.</span>
<span class="sd">    fig_outputfile : Optional[str], optional</span>
<span class="sd">        Path to save plots (RMSE curve and explained variance). If ``None``,</span>
<span class="sd">        plots are saved next to ``output_best_genes_csv``. Default ``None``.</span>
<span class="sd">    max_components : int, optional</span>
<span class="sd">        Maximum number of components to consider in CV. Default ``15``.</span>
<span class="sd">    n_splits : int, optional</span>
<span class="sd">        Number of CV splits. Default ``5``.</span>
<span class="sd">    n_permutations : int, optional</span>
<span class="sd">        Number of permutations for significance testing. Default ``1000``.</span>
<span class="sd">    n_components : Optional[int], optional</span>
<span class="sd">        If provided, forces the number of components; otherwise chosen by CV.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    None</span>
<span class="sd">        Saves outputs to disk and logs progress.</span>

<span class="sd">    Notes</span>
<span class="sd">    -----</span>
<span class="sd">    - Inputs are standardized before fitting.</span>
<span class="sd">    - Significant component selection prefers components with permutation</span>
<span class="sd">      p-values below the threshold; otherwise the component with highest</span>
<span class="sd">      explained variance is chosen.</span>
<span class="sd">    - Output directories are created automatically for CSV and figures.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loading t-values from: </span><span class="si">{</span><span class="n">t_values_file_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loading gene contributions from: </span><span class="si">{</span><span class="n">gene_expression_file_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">df_pairs_values</span><span class="p">,</span> <span class="n">t_values</span><span class="p">,</span> <span class="n">gene_matrix</span><span class="p">,</span> <span class="n">gene_names</span> <span class="o">=</span> <span class="n">read_data_from_files</span><span class="p">(</span><span class="n">t_values_file_path</span><span class="p">,</span> <span class="n">gene_expression_file_path</span><span class="p">)</span>
    <span class="n">df_pairs_values</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_best_genes_path</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;.csv&#39;</span><span class="p">,</span> <span class="s1">&#39;_pairs_values.csv&#39;</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Standardize</span>
    <span class="kn">from</span> <span class="nn">sklearn.preprocessing</span> <span class="kn">import</span> <span class="n">StandardScaler</span>
    <span class="n">scaler</span> <span class="o">=</span> <span class="n">StandardScaler</span><span class="p">()</span>
    <span class="n">X</span> <span class="o">=</span> <span class="n">scaler</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">gene_matrix</span><span class="p">)</span>
    <span class="n">Y</span> <span class="o">=</span> <span class="n">scaler</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">t_values</span><span class="p">)</span>

    <span class="c1"># CV errors and best components</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Selecting optimal components (max=</span><span class="si">{</span><span class="n">max_components</span><span class="si">}</span><span class="s2">, splits=</span><span class="si">{</span><span class="n">cv_splits</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
    <span class="n">best_n_comp</span><span class="p">,</span> <span class="n">mse_scores</span> <span class="o">=</span> <span class="n">select_optimal_components</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">Y</span><span class="p">,</span> <span class="n">max_components</span><span class="o">=</span><span class="n">max_components</span><span class="p">,</span> <span class="n">cv_splits</span><span class="o">=</span><span class="n">cv_splits</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="n">random_state</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Best number of components selected: </span><span class="si">{</span><span class="n">best_n_comp</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># permutation test</span>
    <span class="n">explained_variance_ratio</span><span class="p">,</span> <span class="n">p_values</span><span class="p">,</span> <span class="n">weights</span> <span class="o">=</span> <span class="n">permutation_explained_variance</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">Y</span><span class="p">,</span> <span class="n">best_n_comp</span><span class="p">,</span> <span class="n">n_permutations</span><span class="o">=</span><span class="n">n_permutations</span><span class="p">)</span>
    <span class="n">best_component</span><span class="p">,</span> <span class="n">best_genes</span><span class="p">,</span> <span class="n">z_scores</span><span class="p">,</span> <span class="n">p_values_genes</span> <span class="o">=</span> <span class="n">select_best_component_and_genes</span><span class="p">(</span><span class="n">explained_variance_ratio</span><span class="p">,</span> <span class="n">p_values</span><span class="p">,</span> <span class="n">weights</span><span class="p">)</span>

    <span class="c1"># Save significant genes</span>
    <span class="n">save_best_genes_to_csv</span><span class="p">(</span><span class="n">gene_names</span><span class="p">,</span> <span class="n">best_genes</span><span class="p">,</span> <span class="n">z_scores</span><span class="p">,</span> <span class="n">p_values_genes</span><span class="p">,</span> <span class="n">output_best_genes_path</span><span class="p">)</span>

    <span class="c1"># Plot and save curves and data</span>
    <span class="c1"># Ensure directories exist only when paths are provided</span>
    <span class="k">if</span> <span class="n">mse_plot_path</span><span class="p">:</span>
        <span class="n">mse_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">mse_plot_path</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">mse_dir</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">mse_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># Save MSE data (CV errors per component)</span>
        <span class="n">mse_csv</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">mse_plot_path</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;_data.csv&quot;</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
            <span class="s2">&quot;Components&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">mse_scores</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)),</span>
            <span class="s2">&quot;RMSE&quot;</span><span class="p">:</span> <span class="n">mse_scores</span>
        <span class="p">})</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">mse_csv</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

        <span class="c1"># Plot MSE curve</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">mse_scores</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">mse_scores</span><span class="p">,</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;PLS CV MSE&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Components&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;MSE&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">mse_plot_path</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved SE curve to: </span><span class="si">{</span><span class="n">mse_plot_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved SE data to: </span><span class="si">{</span><span class="n">mse_csv</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">explained_variance_plot_path</span><span class="p">:</span>
        <span class="n">ev_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">explained_variance_plot_path</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">ev_dir</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">ev_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="c1"># Save explained variance data (for X)</span>
        <span class="n">ev_csv</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">explained_variance_plot_path</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;_data.csv&quot;</span>
        <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
            <span class="s2">&quot;Component&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">explained_variance_ratio</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)),</span>
            <span class="s2">&quot;ExplainedVariance&quot;</span><span class="p">:</span> <span class="n">explained_variance_ratio</span>
        <span class="p">})</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">ev_csv</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="c1"># Plot explained variance</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">explained_variance_ratio</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">),</span> <span class="n">explained_variance_ratio</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span>
                 <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;b&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Explained Variance (X)&#39;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;PLS Component&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Explained Variance (%)&quot;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">14</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="s1">&#39;y&#39;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">explained_variance_plot_path</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved explained variance plot to: </span><span class="si">{</span><span class="n">explained_variance_plot_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved explained variance data to: </span><span class="si">{</span><span class="n">ev_csv</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># Ensure output directories exist for CSVs</span>
    <span class="n">out_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">output_best_genes_path</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">out_dir</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">df_best</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
        <span class="s1">&#39;Gene Index&#39;</span><span class="p">:</span> <span class="n">gene_names</span><span class="p">[</span><span class="n">best_genes</span><span class="p">],</span>
        <span class="s1">&#39;Original Z-Score&#39;</span><span class="p">:</span> <span class="n">z_scores</span><span class="p">[</span><span class="n">best_genes</span><span class="p">],</span>
        <span class="s1">&#39;P-Value&#39;</span><span class="p">:</span> <span class="n">p_values_genes</span><span class="p">[</span><span class="n">best_genes</span><span class="p">]</span>
    <span class="p">})</span>
    <span class="n">df_best</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">output_best_genes_path</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Saved best genes CSV to: </span><span class="si">{</span><span class="n">output_best_genes_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Transcriptome PLS pipeline completed successfully.&quot;</span><span class="p">)</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, BrainNetAnno.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>